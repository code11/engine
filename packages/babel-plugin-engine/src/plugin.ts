import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
