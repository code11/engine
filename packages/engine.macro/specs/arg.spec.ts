import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

const macroFile = "'@c11/engine.macro'";
const config = require("./babelConfig.json");

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should support arg": {
      code: `
        import { producer } from ${macroFile}
        const result: producer = ({
          a1 = '123',
          a2 = arg.a1,
          a3 = arg.a2[arg.a1],
          a4 = arg.a3[prop.foo],
        }) => { }
      `,
      snapshot: true,
    },
  },
});
