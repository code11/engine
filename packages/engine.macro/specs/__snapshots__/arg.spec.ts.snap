// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin should support arg: should support arg 1`] = `

import { producer } from '@c11/engine.macro'
const result: producer = ({
  a1 = '123',
  a2 = arg.a1,
  a3 = arg.a2[arg.a1],
  a4 = arg.a3[prop.foo],
}) => { }

      ↓ ↓ ↓ ↓ ↓ ↓

"use strict";

var result = {
  props: {
    type: "STRUCT",
    value: {
      a1: {
        type: "VALUE",
        value: {
          type: "CONST",
          value: "123",
        },
      },
      a2: {
        type: "VALUE",
        value: {
          type: "INTERNAL",
          path: ["a1"],
        },
      },
      a3: {
        type: "VALUE",
        value: {
          type: "INTERNAL",
          path: ["a2", "$a1"],
        },
      },
      a4: {
        type: "VALUE",
        value: {
          type: "INTERNAL",
          path: ["a3", "@foo"],
        },
      },
    },
    meta: {},
  },
  fn: function fn(_ref) {
    var a1 = _ref.a1,
      a2 = _ref.a2,
      a3 = _ref.a3,
      a4 = _ref.a4;
  },
};


`;
